File: tests-cypress/cypress/support/pages/trade/trade-page.ts
import { routes } from '../../constants/routes'
import { testId } from '../../utils/core/selectors'

export const tradeObj = {
  page: testId('trade-page'),
  form: testId('trade-form'),
  marketPrice: testId('trade-market-price'),
  alerts: {
    success: testId('trade-success'),
    error: testId('trade-error'),
  },
  fields: {
    coin: testId('trade-coin'),
    typeBuy: testId('trade-type-buy'),
    typeSell: testId('trade-type-sell'),
    amount: testId('trade-amount'),
    price: testId('trade-price'),
    date: testId('trade-date'),
    note: testId('trade-note'),
    confirm: testId('trade-confirm'),
    submit: testId('trade-submit'),
  },
  errors: {
    coin: testId('trade-coin-error'),
    type: testId('trade-type-error'),
    amount: testId('trade-amount-error'),
    price: testId('trade-price-error'),
    date: testId('trade-date-error'),
    confirm: testId('trade-confirm-error'),
  },
}

export const tradePage = {
  visit: () => {
    cy.visit(routes.trade)
    cy.get(tradeObj.page).should('be.visible')
    cy.get(tradeObj.form).should('be.visible')
  },

  selectFirstCoin: () => {
    cy.get(tradeObj.fields.coin).find('option').should('have.length.greaterThan', 1)
    cy.get(tradeObj.fields.coin)
      .find('option')
      .eq(1)
      .then(option => {
        cy.uiSelect(tradeObj.fields.coin, String(option.val()))
      })
    cy.get(tradeObj.marketPrice).should('be.visible')
  },

  selectType: (type: 'buy' | 'sell') => {
    cy.uiCheck(type === 'buy' ? tradeObj.fields.typeBuy : tradeObj.fields.typeSell)
  },

  enterAmount: (amount: string) => {
    cy.uiType(tradeObj.fields.amount, amount)
  },

  enterPrice: (price: string) => {
    cy.uiType(tradeObj.fields.price, price)
  },

  enterNote: (note: string) => {
    cy.uiType(tradeObj.fields.note, note)
  },

  confirm: () => {
    cy.uiCheck(tradeObj.fields.confirm)
  },

  submit: () => {
    cy.uiClick(tradeObj.fields.submit)
  },

  verifyShellReady: () => {
    cy.get(tradeObj.fields.coin).should('be.visible')
    cy.get(tradeObj.fields.amount).should('be.visible')
    cy.get(tradeObj.fields.submit).should('be.enabled')
  },

  verifyRequiredErrors: () => {
    cy.get(tradeObj.errors.coin).should('be.visible')
    cy.get(tradeObj.errors.amount).should('be.visible')
    cy.get(tradeObj.errors.price).should('be.visible')
    cy.get(tradeObj.errors.confirm).should('be.visible')
  },

  verifySuccess: (text: string) => {
    cy.get(tradeObj.alerts.success).should('be.visible').and('contain.text', text)
  },
}

File: tests-selenium/src/test/java/com/coinpulse/selenium/pages/TradePage.java
package com.coinpulse.selenium.pages;

import com.coinpulse.selenium.constants.Routes;
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.support.ui.Select;

/**
 * Trade page object; selectors mirror the Cypress trade page object.
 */
public class TradePage extends BasePage {

    private static final By PAGE = byTestId("trade-page");
    private static final By FORM = byTestId("trade-form");
    private static final By MARKET_PRICE = byTestId("trade-market-price");
    private static final By SUCCESS = byTestId("trade-success");
    private static final By COIN = byTestId("trade-coin");
    private static final By TYPE_BUY = byTestId("trade-type-buy");
    private static final By TYPE_SELL = byTestId("trade-type-sell");
    private static final By AMOUNT = byTestId("trade-amount");
    private static final By CONFIRM = byTestId("trade-confirm");
    private static final By SUBMIT = byTestId("trade-submit");
    private static final By COIN_ERROR = byTestId("trade-coin-error");
    private static final By AMOUNT_ERROR = byTestId("trade-amount-error");
    private static final By PRICE_ERROR = byTestId("trade-price-error");
    private static final By CONFIRM_ERROR = byTestId("trade-confirm-error");

    private static final String BUY_TYPE = "buy";

    public TradePage(WebDriver driver) {
        super(driver);
    }

    public void open() {
        openPath(Routes.TRADE);
        awaitVisible(PAGE);
        awaitVisible(FORM);
    }

    public void verifyShellReady() {
        awaitVisible(COIN);
        awaitVisible(AMOUNT);
        awaitEnabled(SUBMIT);
    }

    public void selectFirstCoin() {
        Select coinSelect = new Select(awaitVisible(COIN));
        wait.until(d -> coinSelect.getOptions().size() > 1);
        coinSelect.selectByIndex(1);
        awaitVisible(MARKET_PRICE);
    }

    public void selectCoin(long coinId) {
        Select coinSelect = new Select(awaitVisible(COIN));
        wait.until(d -> coinSelect.getOptions().size() > 1);
        coinSelect.selectByValue(String.valueOf(coinId));
        awaitVisible(MARKET_PRICE);
    }

    public void selectType(String type) {
        click(BUY_TYPE.equals(type) ? TYPE_BUY : TYPE_SELL);
    }

    public void enterAmount(String amount) {
        type(AMOUNT, amount);
    }

    public void confirm() {
        click(CONFIRM);
    }

    public void submit() {
        click(SUBMIT);
    }

    public void verifyRequiredErrors() {
        awaitVisible(COIN_ERROR);
        awaitVisible(AMOUNT_ERROR);
        awaitVisible(PRICE_ERROR);
        awaitVisible(CONFIRM_ERROR);
    }

    public void verifySuccess(String text) {
        awaitVisible(SUCCESS);
        awaitTextContains(SUCCESS, text);
    }
}
